Lib/test/test_iter.py

```diff
@@ -238,6 +247,81 @@ def test_mutating_seq_class_exhausted_iter(self):
 
+    def test_reduce_mutating_builtins_iter(self):
+        # This is a reproducer of issue #101765
+        # where iter `__reduce__` calls could lead to a segfault or SystemError
+        # depending on the order of C argument evaluation, which is undefined
+
+        # Backup builtins
+        builtins_dict = builtins.__dict__
+        orig = {"iter": iter, "reversed": reversed}
+
+        def run(builtin_name, item, init=None, sentinel=None):
+            if init is not None:
+                args = init if isinstance(init, tuple) else (init,)
+                item = item(*args)
+            it = iter(item) if sentinel is None else iter(item, sentinel)
+
+            class CustomStr:
+                def __init__(self, name, iterator):
+                    self.name = name
+                    self.iterator = iterator
+                def __hash__(self):
+                    return hash(self.name)
+                def __eq__(self, other):
+                    # Here we exhaust our iterator, possibly changing
+                    # its `it_seq` pointer to NULL
+                    # The `__reduce__` call should correctly get
+                    # the pointers after this call
+                    list(self.iterator)
+                    return other == self.name
+
+            # del is required here
+            # to avoid calling the last test case's custom __eq__
+            del builtins_dict[builtin_name]
+            builtins_dict[CustomStr(builtin_name, it)] = orig[builtin_name]
+
+            return it.__reduce__()
+
+        types = [
+            (EmptyIterClass, ()),
+            (bytes, 8),
+            (bytearray, 8),
+            (tuple, range(8)),
+            (lambda: 0, None, 0),
+            (tuple[int],)  # GenericAlias
+        ]
+
+        try:
+            run_iter = functools.partial(run, "iter")
+            # The returned value of `__reduce__` should not only be valid
+            # but also *empty*, as `it` was exhausted during `__eq__`
+            # i.e "xyz" returns (iter, ("",))
+            self.assertEqual(run_iter(str, "xyz"), (orig["iter"], ("",)))
+            self.assertEqual(run_iter(list, range(8)), (orig["iter"], ([],)))
+
+            # _PyEval_GetBuiltin is also called for `reversed` in a branch of
+            # listiter_reduce_general
+            self.assertEqual(
+                run("reversed", orig["reversed"], list(range(8))),
+                (iter, ([],))
+            )
+
+            for case in types:
+                self.assertEqual(run_iter(*case), (orig["iter"], ((),)))
+        finally:
+            # del is required here
+            # to avoid calling our custom __eq__
+            # we also need to supress KeyErrors in case
+            # a failed test deletes the key without setting anything
+            with contextlib.suppress(KeyError):
+                del builtins_dict["iter"]
+            with contextlib.suppress(KeyError):
+                del builtins_dict["reversed"]
```

Comment:
I'm assuming here that it's important we do both deletions before either set? if not, we could collapse this into the existing for loop

```suggestion
            for key in orig.keys():
                with contextlib.suppress(KeyError):
                    del builtins_dict[key]
```

Reply:
Yes I think this is better behavior wise actually, they can be in the same loop, changed to
```python
for key, func in orig.items():
    with contextlib.suppress(KeyError):
        del builtins_dict[key]
    builtins_dict[key] = func
```